scriptworker/context.py

```diff
@@ -5,114 +5,150 @@
+                     |
+          -----------------------
+          |                     |
+    ScriptContext       BaseWorkerContext
+                                |
+                     -----------------------
+                     |                     |
+               WorkerContext          TaskContext
+
 Attributes:
     log (logging.Logger): the log object for the module.
 
 """
 import aiohttp
-import arrow
 import asyncio
 from copy import deepcopy
-import json
 import logging
 import os
 import tempfile
 
-from scriptworker.utils import makedirs, load_json_or_yaml_from_url
+from scriptworker.utils import load_json_or_yaml_from_url
 from taskcluster.aio import Queue
 
 log = logging.getLogger(__name__)
 
 
-class Context(object):
-    """Basic config holding object.
+# BaseContext {{{1
+class BaseContext(object):
+    """Base context object.
 
-    Avoids putting everything in single monolithic object, but allows for
-    passing around config and easier overriding in tests.
+    Allows for passing around config and easier overriding in tests.
+    Most likely we'll be using non-Base context objects.
 
     Attributes:
         config (dict): the running config.  In production this will be a
             FrozenDict.
-        credentials_timestamp (int): the unix timestamp when we last updated
-            our credentials.
-        proc (task_process.TaskProcess): when launching the script, this is
-            the process object.
-        queue (taskcluster.aio.Queue): the taskcluster Queue object
-            containing the scriptworker credentials.
         session (aiohttp.ClientSession): the default aiohttp session
-        task (dict): the task definition for the current task.
-        temp_queue (taskcluster.aio.Queue): the taskcluster Queue object
-            containing the task-specific temporary credentials.
 
     """
 
     config = None
-    credentials_timestamp = None
-    proc = None
-    queue = None
     session = None
+
+
+# ScriptContext {{{1
+class ScriptContext(BaseContext):
+    """Script context object (e.g. scriptworker.client).
+
+    We'll use this until we transition away from using context objects
+    in our scripts.
+
+    Attributes:
+        task (dict): the running task definition
+
+    """
+
     task = None
-    temp_queue = None
-    running_tasks = None
+
+
+# BaseWorkerContext {{{1
+class BaseWorkerContext(BaseContext):
+    """Base worker context.
+
+    Contains shared structure between both ``WorkerContext`` and
+    ``TaskContext``.
+
+    Scripts shouldn't need to use worker context objects
+
+        queue (taskcluster.aio.Queue): the taskcluster Queue object
+            containing the scriptworker credentials.
+
+    """
+
+    queue = None
     _credentials = None
-    _claim_task = None  # This assumes a single task per worker.
     _event_loop = None
-    _temp_credentials = None  # This assumes a single task per worker.
-    _reclaim_task = None
     _projects = None
```

Comment:
- `queue`: mentioned in my previous comment. We need a temp queue per task for reclaiming/uploading/resolving, and we might as well make it the `TaskContext.queue`.
- `credentials`: similar, but for temp credentials
- `event_loop`: we use the event loop mostly for test purposes, so tests can override which event loop we use rather than having to mock `asyncio.get_event_loop` during integration tests. We likely need access to this, either by specifying it in `TaskContext`, or by also passing in the `WorkerContext` just to access this test-oriented attribute. I think it's cleaner having a copy in `TaskContext`.
- `projects`: We use this to verify CoT, which is task-specific. Again, we can either add it to `TaskContext`, or we can pass in `WorkerContext` as well, either as another function argument, or as `TaskContext.worker_context`. I think the level of copying is minimal enough, and passing around the full `WorkerContext` is going the wrong way.